File: src/application/excel/EquipmentData.java
package application.excel;

import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import org.apache.poi.xssf.usermodel.XSSFCellStyle;
import org.apache.poi.xssf.usermodel.XSSFSheet;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import application.enums.EquipmentType;
import application.models.Armor;
import application.models.Weapon;
import application.models.lists.AbstractItems;
import application.models.lists.Armors;
import application.models.lists.Lists;
import application.models.lists.Shops;
import application.models.lists.Weapons;
import application.staticClasses.TextReader;

public class EquipmentData
{
    public static void makeChangeList()
    {

	try (XSSFWorkbook workbook = new XSSFWorkbook(); OutputStream fileOut = Excel.createOutputStream("Equipment.xlsx"))
	{

	    XSSFSheet weaponsSheet = workbook.createSheet("Weapon Data");
	    XSSFSheet armorSheet = workbook.createSheet("Armor Data");

	    XSSFCellStyle defaultStyle = Excel.setDefaultStyle(workbook);
	    XSSFCellStyle boldStyle = Excel.setBoldStyle(workbook);

	    List<RowData> rowData = fillWeaponSheet(defaultStyle, boldStyle);
	    Excel.writeDataToSheet(weaponsSheet, rowData);
	    rowData = fillArmorSheet(defaultStyle, boldStyle);
	    Excel.writeDataToSheet(armorSheet, rowData);

	    workbook.write(fileOut);
	}
	catch (IOException e)
	{
	    e.printStackTrace();
	}

    }

    private static List<RowData> fillWeaponSheet(XSSFCellStyle defaultStyle, XSSFCellStyle boldStyle)
    {

	Weapons weapons = new Weapons(Lists.getWeapons());

	List<RowData> rowData = new ArrayList<RowData>();

	Weapons axes = new Weapons(Lists.getWeapons().getDModels());
	;
	Weapons knightSwords = new Weapons(weapons.getDModels());
	Weapons swords = new Weapons(weapons.getDModels());
	Weapons knives = new Weapons(weapons.getDModels());
	Weapons rods = new Weapons(weapons.getDModels());
	Weapons fists = new Weapons(weapons.getDModels());
	Weapons misc = new Weapons(weapons.getDModels());

	for (Weapon w : weapons.getModels())
	{
	    switch (EquipmentType.getType(w)) {
		case SWORD:
		    if (w.getEquipCode() == 01)
		    {
			knightSwords.getModels().add(w);
		    }
		    else if ((w.getEquipCode() & 0x08) > 0) // if only Wilme can equip it
		    {
			misc.getModels().add(w);
		    }
		    else
			swords.getModels().add(w);
		    break;
		case AXE:
			axes.getModels().add(w);
		    break;

		case SABER:
		    rods.getModels().add(w);
		    break;
		case ROD:
		    rods.getModels().add(w);
		    break;
		case KNIFE:
		    knives.getModels().add(w);
		    break;
		case HAND:
		    fists.getModels().add(w);
		    break;
		default:
		    misc.getModels().add(w);
		    break;
	    }
	}

	rowData.add(new RowData(defaultStyle, new String[]
	{
		"Name", "Power", "Cost", "Discount", "Equip", "Locations", "Default Locations", "Name Pointer"
	}));
	rowData.add(new RowData(defaultStyle, new String[]
	{
		""
	}));

	Shops shops = new Shops();
	shops.addUsefulModels(Lists.getShops(), false);
	shops.chronologicalIndexSort();

	rowData.add(new RowData(boldStyle, new String[]
	{
		"Knight Swords"
	}));
	addWeaponType(rowData, knightSwords, shops, defaultStyle);

	rowData.add(new RowData(boldStyle, new String[]
	{
		"Swords"
	}));
	addWeaponType(rowData, swords, shops, defaultStyle);

	rowData.add(new RowData(boldStyle, new String[]
	{
		"Axes"
	}));
	addWeaponType(rowData, axes, shops, defaultStyle);

	rowData.add(new RowData(boldStyle, new String[]
	{
		"Knives"
	}));
	addWeaponType(rowData, knives, shops, defaultStyle);
	rowData.add(new RowData(boldStyle, new String[]
	{
		"Rods"
	}));
	addWeaponType(rowData, rods, shops, defaultStyle);
	rowData.add(new RowData(boldStyle, new String[]
	{
		"Fists"
	}));
	addWeaponType(rowData, fists, shops, defaultStyle);
	rowData.add(new RowData(boldStyle, new String[]
	{
		"Misc."
	}));
	addWeaponType(rowData, misc, shops, defaultStyle);
	return rowData;
    }

    private static void addWeaponType(List<RowData> rowData, Weapons weapons, Shops shops, XSSFCellStyle defaultStyle)
    {
	weapons.sortList((Comparator<Weapon>) application.models.lists.Equipment.ascendingComparator);

	for (Weapon w : weapons.getModels())
	{
	    Weapon Dw = weapons.getDModel(w);

	    String namePointer = Excel.CompareValues(TextReader.pointerToText(Dw.getNamePointer()), TextReader.pointerToText(w.getNamePointer()));

	    rowData.add(new RowData(defaultStyle, new String[]
	    {
		    Excel.CompareValues(Dw.getName(), w.getName()), Excel.CompareValues(Dw.getPower(), w.getPower()), Excel.CompareValues(Dw.getCost(), w.getCost()),
		    Excel.CompareValues(Dw.getDiscount(), w.getDiscount()), Excel.CompareValues(getEquipNames(Dw.getEquipCode()), getEquipNames(w.getEquipCode())),
		    AbstractItems.getLocations(shops.getModels(), Lists.getCharacters().getModels(), w.getItemCode()),
		    AbstractItems.getLocations(shops.getDModels(), Lists.getCharacters().getDModels(), Dw.getItemCode()), namePointer
	    }));
	}
	rowData.add(new RowData(defaultStyle, new String[]
	{
		""
	}));
    }

    private static List<RowData> fillArmorSheet(XSSFCellStyle defaultStyle, XSSFCellStyle boldStyle)
    {
	List<RowData> rowData = new ArrayList<RowData>();
	Armors armors = new Armors(Lists.getArmors());
	Armors bodyArmors = new Armors(armors.getDModels());
	Armors robes = new Armors(armors.getDModels());
	Armors coats = new Armors(armors.getDModels());
	Armors miscArmors = new Armors(armors.getDModels());
	Armors shields = new Armors(armors.getDModels());
	Armors accessories = new Armors(armors.getDModels());

	for (Armor a : armors.getModels())
	{
	    switch (EquipmentType.getType(a)) {
		case MAIL:
		    bodyArmors.getModels().add(a);
		    break;
		case ARMOR:
		    bodyArmors.getModels().add(a);
		    break;
		case CLOAK:
		    if (a.getEquipCode() == 8)
		    {
			miscArmors.getModels().add(a);
		    }
		    else
		    {
			robes.getModels().add(a);
		    }
		    break;
		case ROBE:
		    robes.getModels().add(a);
		    break;
		case SHIELD:
		    shields.getModels().add(a);
		    break;
		default:
		    if (a.isCoat())
		    {
			coats.getModels().add(a);
			break;
		    }
		    else if (a.isBodyArmor())
		    {
			miscArmors.getModels().add(a);
			break;
		    }
		    else
		    {
			accessories.getModels().add(a);
			break;
		    }
	    }
	}

	rowData.add(new RowData(defaultStyle, new String[]
	{
		"Name", "Defense", "Cost", "Discount", "Equip", "Lit", "???", "???", "Fire", "Ice", "Vac", "Deb", "Locations", "Default Locations", "Misc."
	}));
	rowData.add(new RowData(defaultStyle, new String[]
	{
		""
	}));

	Shops shops = new Shops(Lists.getShops());
	shops.chronologicalIndexSort();

	rowData.add(new RowData(boldStyle, new String[]
	{
		"Armors"
	}));
	addArmorType(rowData, bodyArmors, shops, defaultStyle);
	rowData.add(new RowData(boldStyle, new String[]
	{
		"Robes"
	}));
	addArmorType(rowData, robes, shops, defaultStyle);
	rowData.add(new RowData(boldStyle, new String[]
	{
		"Coats"
	}));
	addArmorType(rowData, coats, shops, defaultStyle);
	rowData.add(new RowData(boldStyle, new String[]
	{
		"Misc. Armors"
	}));
	addArmorType(rowData, miscArmors, shops, defaultStyle);
	rowData.add(new RowData(boldStyle, new String[]
	{
		"Shields"
	}));
	addArmorType(rowData, shields, shops, defaultStyle);
	rowData.add(new RowData(boldStyle, new String[]
	{
		"Accessories"
	}));
	addArmorType(rowData, accessories, shops, defaultStyle);
	return rowData;
    }

    private static void addArmorType(List<RowData> rowData, Armors armors, Shops shops, XSSFCellStyle defaultStyle)
    {
	armors.sortList((Comparator<Armor>) application.models.lists.Equipment.ascendingComparator);
	for (Armor a : armors.getModels())
	{
	    Armor Da = armors.getDModel(a);

	    String miscData = "";

	    if (a.getNamePointer() != Da.getNamePointer())
	    {
		miscData = "Name Pointer Changed to " + TextReader.pointerToText(a.getNamePointer());
	    }

	    rowData.add(new RowData(defaultStyle, new String[]
	    {
		    Excel.CompareValues(Da.getName(), a.getName()), Excel.CompareValues(Da.getPower(), a.getPower()), Excel.CompareValues(Da.getCost(), a.getCost()),
		    Excel.CompareValues(Da.getDiscount(), a.getDiscount()), Excel.CompareValues(getEquipNames(Da.getEquipCode()), getEquipNames(a.getEquipCode())),
		    Excel.CompareValues(Da.getLaserRes(), a.getLaserRes()), Excel.CompareValues(Da.getUnknownRes1(), a.getUnknownRes1()), Excel.CompareValues(Da.getUnknownRes2(), a.getUnknownRes2()),
		    Excel.CompareValues(Da.getFireRes(), a.getFireRes()), Excel.CompareValues(Da.getIceRes(), a.getIceRes()), Excel.CompareValues(Da.getVacuumRes(), a.getVacuumRes()),
		    Excel.CompareValues(Da.getDebuffRes(), a.getDebuffRes()), AbstractItems.getLocations(shops.getModels(), Lists.getCharacters().getModels(), a.getItemCode()),
		    AbstractItems.getLocations(shops.getDModels(), Lists.getCharacters().getDModels(), Da.getItemCode()), miscData
	    }));
	}
	rowData.add(new RowData(defaultStyle, new String[]
	{
		""
	}));
    }

    public static String getEquipNames(int equipCode)
    {
	StringBuilder sb = new StringBuilder();

	if ((equipCode & 0x01) > 0)
	{
	    sb.append("K");
	}
	if ((equipCode & 0x02) > 0)
	{
	    sb.append("O");
	}
	if ((equipCode & 0x04) > 0)
	{
	    sb.append("E");
	}
	if ((equipCode & 0x08) > 0)
	{
	    sb.append("W");
	}
	if ((equipCode & 0x10) > 0)
	{
	    sb.append("X");
	}
	if ((equipCode & 0x20) > 0)
	{
	    sb.append("V");
	}
	if ((equipCode & 0x40) > 0)
	{
	    sb.append("L");
	}
	return sb.toString();
    }
}

File: src/application/excel/EquipmentData.kt
package application.excel

import application.enums.EquipmentType
import application.models.Armor
import application.models.Weapon
import application.models.lists.*
import application.staticClasses.TextReader
import org.apache.poi.xssf.usermodel.XSSFCellStyle
import java.util.*

class EquipmentData(fileName: String) : Excel(fileName)
{
	override fun makeChangeList()
	{
		val weaponsSheet = workbook.createSheet("Weapon Data")
		val armorSheet = workbook.createSheet("Armor Data")

		weaponsSheet.createFreezePane(2,2,2,2)
		armorSheet.createFreezePane(2,2,2,2)
		writeDataToSheet(weaponsSheet, fillWeaponSheet())
		writeDataToSheet(armorSheet, fillArmorSheet())
	}


	private fun fillWeaponSheet(): SheetData
	{
		val weapons = Weapons(Lists.weapons)
		val axes = Weapons(weapons.dModels)
		val knightSwords = Weapons(weapons.dModels)
		val swords = Weapons(weapons.dModels)
		val knives = Weapons(weapons.dModels)
		val rods = Weapons(weapons.dModels)
		val fists = Weapons(weapons.dModels)
		val misc = Weapons(weapons.dModels)

		val sheetData = SheetData()
		weapons.models.forEach {
			when (EquipmentType.getType(it))
			{
				EquipmentType.SWORD                                         -> if (it.equipCode == 1) knightSwords.models.add(it)
				else if (it.equipCode and 0x08 > 0) misc.models.add(it) // if only Wilme can equip it
				else swords.models.add(it)

				EquipmentType.AXE                                           -> axes.models.add(it)
				EquipmentType.STAFF, EquipmentType.ROD, EquipmentType.SABER -> rods.models.add(it)
				EquipmentType.KNIFE                                         -> knives.models.add(it)
				EquipmentType.HAND                                          -> fists.models.add(it)
				else                                                        -> misc.models.add(it)
			}
		}

		sheetData.addRow(RowData(boldStyle, mutableListOf("Name", "Power", "Cost", "Discount", "Equip", "Locations", "Default Locations", "Name Pointer", "ROM Location")))
		sheetData.addRow(RowData(defaultStyle, mutableListOf("")))

		val shops = Shops()
		shops.addUsefulModels(Lists.shops, false)
		shops.chronologicalIndexSort()

		sheetData.addRow(RowData(boldStyle, mutableListOf("Knight Swords")))
		addWeaponType(sheetData, knightSwords, shops, defaultStyle)

		sheetData.addRow(RowData(boldStyle, mutableListOf("Axes")))
		addWeaponType(sheetData, axes, shops, defaultStyle)

		sheetData.addRow(RowData(boldStyle, mutableListOf("Swords")))
		addWeaponType(sheetData, swords, shops, defaultStyle)
		
		sheetData.addRow(RowData(boldStyle, mutableListOf("Knives")))
		addWeaponType(sheetData, knives, shops, defaultStyle)

		sheetData.addRow(RowData(boldStyle, mutableListOf("Rods")))
		addWeaponType(sheetData, rods, shops, defaultStyle)

		sheetData.addRow(RowData(boldStyle, mutableListOf("Fists")))
		addWeaponType(sheetData, fists, shops, defaultStyle)

		sheetData.addRow(RowData(boldStyle, mutableListOf("Misc.")))
		addWeaponType(sheetData, misc, shops, defaultStyle)
		return sheetData
	}

	private fun addWeaponType(rowData: SheetData, weapons: Weapons, shops: Shops, defaultStyle: XSSFCellStyle)
	{
		weapons.sortList(application.models.lists.Equipment.ascendingComparator as Comparator<Weapon>)

		weapons.models.forEach {
			val Dw = weapons.getDModel(it)
			val namePointer = CompareValues(TextReader.pointerToText(Dw.namePointer), TextReader.pointerToText(it.namePointer))
			rowData.addRow(RowData(defaultStyle,
				mutableListOf(CompareValues(Dw.name, it.name),
					CompareValues(Dw.power, it.power),
					CompareValues(Dw.cost, it.cost),
					CompareValues(Dw.discount, it.discount),
					CompareValues(getEquipNames(Dw.equipCode),
						getEquipNames(it.equipCode)),
					AbstractItems.getLocations(shops.models,
						Lists.characters.models, it.itemCode),
					AbstractItems.getLocations(shops.dModels,
						Lists.characters.dModels, Dw.itemCode),
					namePointer,
					"0x${Integer.toHexString(it.offset).toUpperCase()}")))
		}
		rowData.addRow(RowData(defaultStyle, mutableListOf("")))
	}

	private fun fillArmorSheet(): SheetData
	{
		val sheetData = SheetData()
		val armors = Armors(Lists.armors)
		armors.models.removeIf{it.equipCode==0}


		val bodyArmors = Armors(armors.dModels)
		val robes = Armors(armors.dModels)
		val coats = Armors(armors.dModels)
		val miscArmors = Armors(armors.dModels)
		val shields = Armors(armors.dModels)
		val accessories = Armors(armors.dModels)

		armors.models.forEach {
			when (EquipmentType.getType(it))
			{
				EquipmentType.MAIL   -> bodyArmors.models.add(it)
				EquipmentType.ARMOR  -> bodyArmors.models.add(it)
				EquipmentType.CLOAK  -> if (it.equipCode == 8) miscArmors.models.add(it)
				else robes.models.add(it)
				EquipmentType.ROBE   -> robes.models.add(it)
				EquipmentType.SHIELD -> shields.models.add(it)
				else                 -> if (it.isCoat) //Default
					coats.models.add(it)
				else if (it.isBodyArmor) miscArmors.models.add(it)
				else accessories.models.add(it)
			}
		}

		sheetData.addRow(RowData(boldStyle, mutableListOf("Name", "Defense", "Cost", "Discount", "Equip", "Lit", "???", "???", "Fire", "Ice", "Vac", "Deb", "Locations", "Default Locations", "Name Pointer", "ROM Location")))
		sheetData.addRow(RowData(defaultStyle, mutableListOf("")))

		val shops = Shops(Lists.shops)
		shops.chronologicalIndexSort()

		sheetData.addRow(RowData(boldStyle, mutableListOf("Armors")))
		addArmorType(sheetData, bodyArmors, shops, defaultStyle)
		sheetData.addRow(RowData(boldStyle, mutableListOf("Robes")))
		addArmorType(sheetData, robes, shops, defaultStyle)
		sheetData.addRow(RowData(boldStyle, mutableListOf("Coats")))
		addArmorType(sheetData, coats, shops, defaultStyle)
		sheetData.addRow(RowData(boldStyle, mutableListOf("Misc. Armors")))
		addArmorType(sheetData, miscArmors, shops, defaultStyle)
		sheetData.addRow(RowData(boldStyle, mutableListOf("Shields")))
		addArmorType(sheetData, shields, shops, defaultStyle)
		sheetData.addRow(RowData(boldStyle, mutableListOf("Accessories")))
		addArmorType(sheetData, accessories, shops, defaultStyle)
		return sheetData
	}

	private fun addArmorType(rowData: SheetData, armors: Armors, shops: Shops, defaultStyle: XSSFCellStyle)
	{
		armors.sortList(application.models.lists.Equipment.ascendingComparator as Comparator<Armor>)
		armors.models.forEach {
			val Da = armors.getDModel(it)



			val namePointer = CompareValues(TextReader.pointerToText(Da.namePointer), TextReader.pointerToText(it.namePointer))
			rowData.addRow(RowData(defaultStyle, mutableListOf(CompareValues(Da.name, it.name), CompareValues(Da.power, it.power), CompareValues(Da.cost, it.cost), CompareValues(Da.discount, it.discount), CompareValues(getEquipNames(Da.equipCode), getEquipNames(it.equipCode)), CompareValues(Da.laserRes, it.laserRes), CompareValues(Da.unknownRes1, it.unknownRes1), CompareValues(Da.unknownRes2, it.unknownRes2), CompareValues(Da.fireRes, it.fireRes), CompareValues(Da.iceRes, it.iceRes), CompareValues(Da.vacuumRes, it.vacuumRes), CompareValues(Da.debuffRes, it.debuffRes), AbstractItems.getLocations(shops.models, Lists.characters.models, it.itemCode), AbstractItems.getLocations(shops.dModels, Lists.characters.dModels, Da.itemCode), namePointer, "0x${Integer.toHexString(it.offset).toUpperCase()}")))
		}
		rowData.addRow(RowData(defaultStyle, mutableListOf("")))
	}

	fun getEquipNames(equipCode: Int): String
	{
		val sb = StringBuilder()
		if (equipCode and 0x01 > 0) sb.append("K")
		if (equipCode and 0x02 > 0) sb.append("O")
		if (equipCode and 0x04 > 0) sb.append("E")
		if (equipCode and 0x08 > 0) sb.append("W")
		if (equipCode and 0x10 > 0) sb.append("X")
		if (equipCode and 0x20 > 0) sb.append("V")
		if (equipCode and 0x40 > 0) sb.append("L")
		return sb.toString()
	}
}
